In [3]:
import pandas as pd
import os
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [15]:
q_iso=pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-i-karamnov-39/data.csv', encoding='Windows-1251' )

In [16]:
q_iso

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [17]:
q_iso=q_iso.drop_duplicates()

In [18]:
q_iso.query('Quantity > 0')

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[526054 rows x 8 columns]

💀 Сложное задание! 💀

Посчитайте число заказов для каждого пользователя (CustomerID) из Германии (Germany). Оставьте только тех, кто совершил более N транзакций (InvoiceNo), где N – 80% процентиль. Запишите полученные id пользователей в germany_top (не весь датафрейм, только id).

Отфильтрованные данные записаны в retail. Идентификатор заказа – InvoiceNo. Для каждого заказа в данных может встречаться более 1 строки.

Подсказки

- **Как решать это задание?**
    
    Отберите клиентов только из Германии (`Country`), сгруппируйте по клиентам (`CustomerID`) и посчитайте число заказов для каждого клиента. Затем на полученном результате используйте метод `quantile`, чтобы получить данные по 80-ому перцентилю. После этого отфильтруйте Германских клиентов, оставив с числом заказов больше чем 80-ый перцентиль. Запишите `CustomerID` полученных клиентов в переменную `germany_top`
    
- **У меня не сходится ответ!**
    
    А какую функцию вы применяли при агрегации после группировки по клиентам? Там нужно посчитать число заказов, при том, что на 1 заказ может быть несколько строк - например, заказ получен, заказ в процессе, заказ доставлен. Поэтому в агрегации нужно использовать функцию `'nunique'` c колонкой `InvoiceNo`.
    
- **Подсчёт числа уникальных значений**
    
    Используйте метод `nunique`, чтобы получить число уникальных значений для каждой группы. В агрегации его можно указать как метод `pd.Series.nunique` либо как строку `'nunique'` (такой вариант работает быстрее)
    
- **Как взять 80-ый перцентиль?**
    
    Воспользуйтесь методом `quantile`, указав внутри него какой перцентиль вам нужен в виде дроби. К примеру
    
    ```python
    df.column1.quantile(0.5)
    ```
    
    даст вам 50-ый перцентиль по значениям в колонке `column1`
    
- **Как использовать переменную внутри** `query`**?**
    
    Чтобы сослаться на переменную внутри `query`, напишите её в запросе и поставьте перед ней `@`. Например
    
    ```python
    threshold = 3
    df.query('col1 >= @threshold')
    ```
    
    отберёт из датафрэйма строки, где в колонке `col1` находятся значения больше либо равные 3
    
- **Я использовал** `groupby`**, не указывая** `as_index=False`**, и** `CustomerID` **оказались в индексе, как их достать?**
    
    Самый простой вариант - обратитесь к атрибуту `index` у нужного датафрэйма. То есть
    
    ```python
    df.index
    ```
    
    Вернёт вам индекс датафрэйма `df`

In [19]:
count_germany=q_iso.query('Country == "Germany"')\
                    .groupby('CustomerID', as_index=False)\
                    .agg({'InvoiceNo' : 'nunique'})
quant=count_germany.InvoiceNo.quantile(q=0.8)
count_germany=count_germany.query('InvoiceNo > @quant')
germany_top=count_germany.CustomerID

In [20]:
#q_iso.groupby('Country').agg({'CustomerID' : 'count'})

In [21]:
germany_top

3     12471.0
4     12472.0
6     12474.0
8     12476.0
12    12481.0
16    12500.0
37    12569.0
44    12600.0
52    12619.0
53    12621.0
57    12626.0
65    12647.0
70    12662.0
78    12705.0
79    12708.0
80    12709.0
82    12712.0
84    12720.0
Name: CustomerID, dtype: float64

⭐️ Задание со звёздочкой! ⭐️

Используя объект с id пользователей (germany_top), полученный на предыдущем шаге, отфильтруйте наблюдения и оставьте в данных записи только по интересующим нас юзерам. Результирующий датафрейм запишите в top_retail_germany.

Датафрейм записан в retail, нужные id пользователей – в germany_top
- **Как делать это задание?**
    
    Отберите из датафрэйма только отобранных раннее Германских пользователей (хранятся в переменной `germany_top`) и поместите сабсэт в переменную `top_retail_germany`
    
- **Как отобрать строки, значения в которых входят в какой-то набор?**
    
    Воспользуйтесь `query` и используйте там оператор `in`, чтобы отобрать нужные строки, то есть этот код
    
    ```python
    vals = [1, 3, 12]
    
    df.query('col1 in @vals')
    ```
    
    отберёт из датафрэйма строки, где в колонке `col1` находятся значения из списка `vals` (1, 3 или 12)

In [22]:
top_retail_germany=q_iso.loc[q_iso.CustomerID.isin(germany_top)]

In [23]:
top_retail_germany=top_retail_germany.drop_duplicates()

Сгруппируйте top_retail_germany по коду товара (StockCode). Какой из продуктов добавляли в корзину чаще всего, кроме POST? 

Note: одним заказом считается единовременная покупка любого количества товара, т.е. без учета Quantity.
- **Как решать это задание?**
    
    Произведите группировку по колонке `StockCode` и посчитайте для каждого кода число раз, когда он был добавлен в корзину. Затем возьмите код товара, который идёт до кода `POST`
    
- **Был добавлен в корзину? Что это вообще значит?**
    
    Имеется в виду число строк с ним - каждое добавление товара в корзину отображено в датафрэйме как строка с кодом этого товара
    
- **Как взять товар до POST?**
    
    Есть разные варианты, самый простой - отсортируйте датафрэйм и возьмите идущий перед `POST`
    
- **Сортировка датафрэйма**
    
    Для сортировки датафрэйма используется метод `sort_values`. Он принимает название колонки, по которой будет идти сортировка. К примеру,
    
    ```python
    df.groupby('pickup_month') \
    .agg({'temp': sum}) \
    .sort_values('temp')
    ```
    
    После подсчёта суммы по температур по месяцу сортирует строки по получившейся колонке (сумме температур, которая называется `temp`)

In [24]:
top_retail_germany.groupby('StockCode').agg({'InvoiceNo':'count'}).sort_values('InvoiceNo')

InvoiceNo
StockCode           
22548              1
22948              1
84800L             1
22956              1
22174              1
...              ...
22467             37
22328             45
22423             55
22326             62
POST             219

[1168 rows x 1 columns]

Вернемся к анализу полного датасета retail. Создайте колонку Revenue с суммой покупки, используя колонки Quantity и UnitPrice.

- **Как посчитать сумму покупки?**
    
    Для этого нужно умножить количество купленного товара на его цену, например
    
    Коржики стоят по 7 рублей
    
    Я взял 8 коржиков
    
    Сумма покупки будет
    
    8 штук * 7 рублей == 56 рублей
    
- **Создание колонки**
    
    Чтобы сделать новую колонку, достаточно написать такой код
    
    ```python
    df['new_column'] = some_value
    ```
    
    где
    
    `df` - датафрэйм
    
    `new_column` - название новой колонки
    
    `some_value` - то, чему будет равна колонка (это может быть одно число или колонка датафрэйма, где число элементов такое же как число строк в `df`)

In [25]:
q_iso[['Quantity', 'UnitPrice']] = q_iso[['Quantity', 'UnitPrice']]. fillna (0)

/opt/tljh/user/lib/python3.7/site-packages/pandas/core/frame.py:3065: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


In [30]:
q_iso.assign(ctr=q_iso.Quantity * q_iso.UnitPrice)

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  Revenue    ctr  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom    15.30  15.30  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34  20.34  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom    22.00  22.00  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34  20.34  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34  20.34  
...                 ...        ...         ...             ...      ...    ...  
541904  12/9/2011 12:50       0.85     12680.0          France    10.20  10.20  
541905  12/9/2011 12:50       2.10     12680.0          France    12.60  12.60  
541906  12/9/2011 12:50       4.15     12680.0          France    16.60  16.60  
541907  12/9/2011 12:50       4.15     12680.0          France    16.60  16.60  
541908  12/9/2011 12:50       4.95     12680.0          France    14.85  14.85  

[536641 rows x 10 columns]

In [31]:
q_iso = q_iso.assign(Revenue=q_iso.Quantity * q_iso.UnitPrice)

Для каждой транзакции (InvoiceNo), посчитайте финальную сумму заказа. В качестве ответа укажите топ-5 (InvoiceNo) по сумме заказа (через запятую с пробелом и в порядке убывания TotalRevenue)

Например, для следующего примера...

InvoiceNo StockCode Quantity InvoiceDate          UnitPrice  CustomerID  Revenue
536365    85123A    6        2010-12-01 08:26:00  2.55       17850.0     15.30
536365    71053     6        2010-12-01 08:26:00  3.39       17850.0     20.34
555555    71053     1        2010-12-02 18:00:00  3.39       17850.0     3.39


...сумма заказа будет равна:

InvoiceNo  TotalRevenue
536365     35.64
555555     3.39


И ответом будет:

536365, 555555

- **Как делать это задание?**
    
    Сгруппируйте данные по номерам транзакций (`InvoiceNo`) и посчитайте сумму по колонке `Revenue`. Отсортируйте результат, и возьмите 5 номеров транзакций с максимальной суммой прибыли
    
- **Мой ответ не принимается!**
    
    А вы указываете номера транзакций от транзакции с наибольшей прибылью к транзакции с наименьшей?
    
- **Как отформатировать ответ не вручную?**
    
    Об этом в рассказывается конце разбора

In [39]:
q_iso.groupby('InvoiceNo').agg({'Revenue' : 'sum'}).sort_values('Revenue', ascending=False)

Revenue
InvoiceNo           
581483     168469.60
541431      77183.60
574941      52940.94
576365      50653.91
556444      38970.00
...              ...
C580605    -17836.46
C550456    -22998.40
C556445    -38970.00
C541433    -77183.60
C581484   -168469.60

[25900 rows x 1 columns]

In [ ]:
581483, 541431, 574941, 576365, 556444